Initial two-stage user provisioning modeling, tests, and import flow

czue opened the pull request:
Review by commit.

##### SUMMARY

For some COVID projects coming up we will likely need a more "sensible" mobile worker provisioning model than "the admin chooses your password". This is a first pass at modeling how that might work.

Much more context in [the working spec doc](https://docs.google.com/document/d/1gZlOGroRTeUMHXwzCuOVw2EUuJcjOsAaAe1qc2iUEe0/edit)

Basically the idea is to add a way to create "non-provisioned" mobile workers who are default deactivated upon creation, but then have a way to set a password and activate their account via email. There's a fair amount of justification for using *this* model in the Google doc - to minimize surface area of the change and maximize re-use of existing paradigms.

Putting this up for review now because:

1. Due to the nature/timelines of the work it's quite time sensitive
2. I want to get any feedback on the modeling/planning side as early as possible to minimize the need to rework things later

cc-ing a hastily and arbitrarily chosen list of potentially interested parties: @orangejenny @dannyroberts @ctsims @snopoke @AndreaMKing5 @esoergel @kaapstorm @biyeun 

##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->

This will initially be rolled out under a new Solutions "limited use" flag until we better understand the product requirements, though has the potential to be a valuable core feature down the line.

##### PRODUCT DESCRIPTION
<!--- For non-invisible changes, describe user-facing effects. -->

See spec for some thoughts on this, though not 100% finalized.

czue committed:
add is_provisioned to CommCareUser model with explanation of intention

czue committed:
add feature flag

czue committed:
cleanup: replace old docstring with slighlty more helpful docstring

czue committed:
cleanup: add basic test for normal user create functionality

czue committed:
allow passing is_provisioned when creating mobile worker

- also set is_active to false if not provisioned
- and add a simple test case

czue committed:
fix setting of is_active on the django user model

also adds test revealing this issue

czue committed:
add test case for is_active explicitly overriding is_provisioned

orangejenny commented:
@czue Haven't thoroughly read the code or the spec yet, but this sounds vaguely like mobile worker self-registration, wanted to confirm you're aware of that feature in case there's any useful overlap.

czue commented:
> this sounds vaguely like mobile worker self-registration, wanted to confirm you're aware of that feature in case there's any useful overlap.

Cool thanks! I'm not familiar with that feature. If you have a pointer to docs or anything that'd be awesome - else I can go digging through wikis/code.

orangejenny commented:
@czue Wiki turned up [this doc](https://confluence.dimagi.com/display/commcarepublic/SMS+Self+Registration), which unfortunately says it's not available on CC versions higher than 2.44. I don't have context on why that is. If you want to talk to a human, @AndreaMKing5 is probably the best person to start with.

ctsims commented:
Context on 2.44:

Self Registration had a mobile app specific component which sent a coded
SMS to the phone which had the app install target in a base64 encoded
message, and the app used to be able to find those messages and
automatically install your app based on the contents. We had to turn that
off in future app releases because Google locked down what apps could
request access to SMS messages.

-Clayton

On Mon, Mar 16, 2020 at 10:56 AM Jenny Schweers <notifications@github.com>
wrote:

> @czue <https://github.com/czue> Wiki turned up this doc
> <https://confluence.dimagi.com/display/commcarepublic/SMS+Self+Registration>,
> which unfortunately says it's not available on CC versions higher than
> 2.44. I don't have context on why that is. If you want to talk to a human,
> @AndreaMKing5 <https://github.com/AndreaMKing5> is probably the best
> person to start with.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/26864#issuecomment-599582057>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABF3OQAY5JNOR2Z5W6RY3LRHY42HANCNFSM4LMJBPUA>
> .
>

snopoke commented:
Looks OK to me.

orangejenny commented on corehq/apps/users/tests/test_create_mobile_workers.py:
```diff
@@ -47,11 +47,27 @@ def test_create_unprovisioned(self):
         self.assertEqual(False, django_user.is_active)
         self.assertEqual(False, self.client.login(username='mw1', password='s3cr4t'))
+
+    def test_is_active_overrides_is_provisioned(self):
```
What would you think of putting an assertion inside of `create` to verify `is_provisioned or not is_active`?

czue replied:
You're essentially saying to disallow this case and update the test accordingly, yeah?

That makes sense to me. Will change.

czue replied:
https://github.com/dimagi/commcare-hq/pull/26864/commits/fc0e6647a64dca27db8ad2bde489e41db299080b

orangejenny reviewed (approved):
Looks good to me, left one suggestion.

czue committed:
make it illegal to create active, unprovisioned workers

czue committed:
rename is_provisioned --> is_account_confirmed for clarity

czue committed:
rename test

czue committed:
add failing test for bulk import updating is_account_confirmed

czue committed:
set is_account_confirmed if present in upload

czue committed:
extract string to boolean or none helper function to its own module

czue committed:
better error message, and allow new field in UI

czue committed:
add validator for new column (small refactor) and update tests

czue committed:
don't validate for required passwords if account isn't confirmed

czue committed:
feed the stickler (who actually had a good suggestion)

czue commented:
@orangejenny @snopoke I renamed the field to `is_account_confirmed` and added support for setting it via bulk upload. This feels like a pretty good stopping point from a "unit of work" perspective, though I plan to continue adding to this and the feature is still "incomplete" - in that an unconfirmed account is just the same as an inactive account right now.

Do you have any preference between merging in a bunch of partial PRs like this one versus some other workflow (e.g. "approve" this one, and then do PRs into this branch until it's at an acceptable V1 and then merge that big thing into master)?

czue commented:
> @czue Wiki turned up [this doc](https://confluence.dimagi.com/display/commcarepublic/SMS+Self+Registration), which unfortunately says it's not available on CC versions higher than 2.44. I don't have context on why that is. If you want to talk to a human, @AndreaMKing5 is probably the best person to start with.

Thanks for sharing this. I took a look through the doc and the code and I think there's probably not a ton of overlap (the SMS workflow creates fully active users and allows people to self-register which is a bit different from the use case we want here where the users are still centrally managed, but need to complete a step before being active). But I did find a few utilities like `random_password()` that will likely be helpful.

orangejenny commented on corehq/apps/users/models.py:
```diff
@@ -1645,7 +1645,7 @@ class CommCareUser(CouchUser, SingleMembershipMixin, CommCareMobileContactMixin)
     # used by user provisioning workflow. defaults to true unless explicitly overridden during
     # user creation
-    is_provisioned = BooleanProperty(default=True)
+    is_account_confirmed = BooleanProperty(default=True)
```
Good rename.

orangejenny commented:
@czue I'd slightly prefer to merge this and continue on with small PRs, to make review easier.

I could imagine wanting the new flag to show up in the download in the future, and then modifying the validation so you can smoothly download and re-upload the same file (probably just modifying it so `is_account_confirmed` can be false in the upload as long as that's not changing the value). But I think it's fine for now.

czue commented:
> @czue I'd slightly prefer to merge this and continue on with small PRs, to make review easier.

Great, this is also my preference.

> I could imagine wanting the new flag to show up in the download in the future, and then modifying 
the validation so you can smoothly download and re-upload the same file (probably just modifying it so `is_account_confirmed` can be false in the upload as long as that's not changing the value). But I think it's fine for now.

Agreed this makes good sense to me. I think currently you can't export deactivated users (at least that I could find?) so it also feels less pressing to me.

orangejenny commented:
> I think currently you can't export deactivated users (at least that I could find?)

Cool, I didn't know that. Agree that makes it low priority to put this in the download.

snopoke commented on corehq/apps/user_importer/importer.py:
```diff
@@ -335,11 +335,21 @@ def create_or_update_users_and_groups(domain, user_specs, group_memoizer=None, u
 
+                    # note: explicitly not including "None" here because that's the default value if not set.
+                    # False means it was set explicitly to that value
+                    if is_account_confirmed is not False:
```
a bit confused by this. This is under the branch for updating existing users but the error message appears to be for new users. Also won't this error if you leave the column blank?

czue replied:
If you leave the column blank I'm pretty sure it will be `None` (I thought there was a test for this - will double check/confirm tomorrow).

But you're right this should be `is False`. I broke it (and the tests) on the last stickler commit. fixing now. https://github.com/dimagi/commcare-hq/pull/26864/commits/60cbfe38c5631e5a1ea326392379aff7cb1aac9c

czue replied:
https://github.com/dimagi/commcare-hq/pull/26864/commits/64e1d9de6f31bef106e4122d2fe10c47b8de5f46

czue replied:
confirmed that leaving the column blank works fine but there's not a test. I'll add it in the next PR

czue replied:
https://github.com/dimagi/commcare-hq/pull/26888

snopoke commented on corehq/apps/user_importer/validation.py:
```diff
@@ -171,7 +181,10 @@ class NewUserPasswordValidator(ImportValidator):
+
+        # explicitly check is_account_confirmed against False because None is the default
+        if not user_id and not is_password(password) and is_account_confirmed is not False:
             return self.error_message
```
should this error message change?

czue replied:
Do you mean like, change to ""New users must have a password set *unless is_account_confirmed is False*."?

I think that's probably an unnecessary detail to include - particularly at this stage and given that this workflow will be behind a feature flag for a while.

(In case it's not clear, all this does is *allow* new users to be created without a password if `is_account_confirmed == False`. Previously it was never allowed.)

czue replied:
https://github.com/dimagi/commcare-hq/pull/26864/commits/64e1d9de6f31bef106e4122d2fe10c47b8de5f46

czue committed:
fix bad stickler change

esoergel reviewed (commented):
:+1:

czue commented on corehq/apps/user_importer/importer.py:
```diff
@@ -319,9 +320,8 @@ def create_or_update_users_and_groups(domain, user_specs, group_memoizer=None, u
-                if is_active and isinstance(is_active, str):
-                    is_active = string_to_boolean(is_active)
+                is_active = spec_value_to_boolean_or_none(row, 'is_active')
+                is_account_confirmed = spec_value_to_boolean_or_none(row, 'is_account_confirmed')
```
This is a good suggestion for when orgs want to operate this way. Knowing how "messy" things often get in the real world I'd be hesitant to not design for the flexible version though. So I might lean to making it a first-class project setting that's configurable behind the toggle and probably lives alongside "strong mobile worker passwords" in the domain settings UI, so orgs have the choice of global versus individual enforcement.

esoergel replied:
That's a reasonable position to take.  Fine by me.

AndreaMKing5 commented:
@czue nice work on the documentation that goes along with this!
